src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -55,19 +55,30 @@ private static void FreeWriteBuffer(Task t, object buffer)
             t.GetAwaiter().GetResult();
         }
 
-        // If we have a read buffer from the pool, return it.
-        private void FreeReadBuffer()
+        //We will only free the read buffer if it
+        //actually contains no decrypted or encrypted bytes
+        private bool TryFreeReadBuffer()
         {
-            if (_internalBuffer != null)
+            if (_internalBuffer != null && _decryptedBytesCount == 0 && _internalBufferCount == 0)
             {
                 ArrayPool<byte>.Shared.Return(_internalBuffer);
                 _internalBuffer = null;
+                _internalBufferCount = 0;
+                _internalOffset = 0;
+                _decryptedBytesCount = 0;
+                _decryptedBytesOffset = 0;
+                return true;
             }
+            return false;
```

Comment:
Why does it return a bool if none of the callers use the bool?

Reply:
No good reason, just the Tryxxx convention but it didn't sit well with me either so I removed it.

Reply:
Doesn't need to be named `TryXxx`... could, for example, be something like `ReturnBufferIfEmpty`.

Reply:
I like that a lot better, renamed.